docs/when_internals.md

```diff
@@ -0,0 +1,342 @@
+        corresponding port name
+      * `_default_drivers`: used to track the default driver for values
+      * `_is_when_builder_`: True (used externally to find when builders)
+    * Properties
+      * `default_drivers`, `input_to_index`, `output_to_index`, `block`: public
+        APIs to internal attributes
+    * Methods
+      * `_check_existing_derived_ref`: if value is a child of an array or tuple
+        that has already been added, we return the child of the existing value,
+        rather than adding a new port, which allows us to maintain bulk
+        assignments in the eventual generated if statement, rather that
+        elaborating into per-child assignments
+      * `_generic_add`: common logic for `add_drivee`, `add_driver`, and `add_condition` which
+        updates the appropriate maps, adds/gets a corresponding builder port,
+        and wires the port to the value
+      * `add_drivee`, `add_driver`, `add_condition`: uses `_generic_add` with
+        specific maps for each variant
+      * `add_default_driver`: adds ports for the driver and drivee, updates the
+        `_default_drivers` map
+      * `remove_default_driver`: removes drivee from `_default_drivers` map
+        (used for unwire)
+      * `_finalize`: run latch inference, add default logic for memories and
+        registers.
+  * `is_when_builder`: public API for checking if instance of when builder
+
+* `protocol_type.py`
+  * `Protocol` should pass the `set_enclosing_when_context` API through to the
+    underlying magma value
+* `util.py`
+  * `reset_global_context`, reset the when context global state (used for
+    teardown of tests that could potentially error and leave a when context in
+    place)
+* `when.py`
+  * `WhenSyntaxError`: used to report errors about invalid ordering of when
+    statements (e.g. an `otherwise` must follow a `when` or `elsewhen`)
+  * `ConditionalWire`: a container with fields `drivee` and `driver` used
+    to store information about conditional wires
+  * `_BlockBase`
+    * Overview
+      * Base class containing common logic for the various types of when blocks
+        (i.e. when, elsewhen, otherwise)
+    * Attributes:
+      * `self._parent`: for `WhenBlock`, reference to enclosing `WhenBlock` if
+        it exists.  For all other blocks, reference to intial `WhenBlock`
+      * `self._children`: for `WhenBlock`, reference to child blocks created by
+        the `spawn` API
+      * `self._conditional_wires`: list of conditional wires created in this block
+      * `self._default_drivers`: list of `_default_drivers` for conditional
+        values
+    * Properties
+      * `root`: traverse `_parent` chain until finding the root
+      * `condition`: get condition for this block
+    * Methods
+      * `spawn`: create a child `_WhenBlock`
+      * `get_conditional_wires_for_drivee`: return all conditional wires corresponding
+        to a drivee
```

Comment:
I think there should only be one conditional wire per drivee per context, so I think this and all consuming code could be simplified to assume/assert a single driver.